scripts/freeu.py

```diff
@@ -9,65 +10,147 @@
 txt2img_steps_callbacks = []
 img2img_steps_callbacks = []
 
+user_options = []
+freeu_options = []
+all_options = []
 
 class FreeUScript(scripts.Script):
     def title(self):
         return "FreeU"
 
     def show(self, is_img2img):
         return scripts.AlwaysVisible
+    
+    def load_function(self):
+        path = os.path.join(os.path.dirname(__file__), "user-custom-settings.json")
+        if os.path.exists(path): # check if file exists
+            with open(path, "r") as f:
+                all_options = json.load(f)
+        else: # if file does not exist, return empty array
+            all_options = []
+        return all_options
+
+    def save_function(self, options):
+        path = os.path.join(os.path.dirname(__file__), "user-custom-settings.json")
+        with open(path, "w") as f:
+            json.dump(options, f)
+
+    def get_index_by_name(self, name: str, array: list) -> int:
+        for i in range(len(array)):
+            if array[i]["name"] == name:
+                return i
+        return -1
 
     def ui(self, is_img2img):
+
+        default_stage_infos = [
+            global_state.StageInfo(1.2, 0.9),
+            global_state.StageInfo(1.4, 0.2),
+            global_state.StageInfo(1, 1),
+        ]
+        sd21_default_stage_infos = [
+            global_state.StageInfo(1.1, 0.9),
+            global_state.StageInfo(1.2, 0.2),
+            global_state.StageInfo(1, 1),
+        ]
+        sdxl_default_stage_infos = [
+            global_state.StageInfo(1.1, 0.6),
+            global_state.StageInfo(1.2, 0.4),
+            global_state.StageInfo(1, 1),
+        ]
+
+        user_options.clear()
+        freeu_options.clear()
+        all_options.clear()
+
+        user_options.extend(self.load_function())
+        freeu_options.extend([
+            {"name": "SD1.4 Recommendations", "freeu": [
+                v
+                for stage_info in default_stage_infos
+                for v in stage_info.to_dict(include_default=True).values()
+            ], "SystemDefault": True, "start": 0.2, "stop": 1, "smooth":0},
+            {"name": "SD2.1 Recommendations", "freeu": [
+                v
+                for stage_info in sd21_default_stage_infos
+                for v in stage_info.to_dict(include_default=True).values()
+            ], "SystemDefault": True, "start": 0.2, "stop": 1, "smooth":0},
+            {"name": "SDXL Recommendations", "freeu": [
+                v
+                for stage_info in sdxl_default_stage_infos
+                for v in stage_info.to_dict(include_default=True).values()
+            ], "SystemDefault": True, "start": 0.2, "stop": 1, "smooth":0},
+            ])
+        
+        all_options.extend(freeu_options)
+        all_options.extend(user_options)
+
         with gr.Accordion(open=False, label=self.title()):
-            with gr.Row():
+            with gr.Row(margin="0.5em"):
```

Comment:
I forgot to test it. Something somewhere said it adds margins between buttons... didn't work :(